### PR TITLE
SWORD: if custom terms disabled, report error #8580

### LINKED ISSUE #8580: SWORD API: Submitting a dataset with terms when custom dataset terms is disabled results in nondescript command line error and null ptr log error.
When attempting to create a dataset using the SWORD api with rights/terms defined in the atom file, the command fails with a generic error and with a null ptr server log entry:

<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<sword:error xmlns="http://www.w3.org/2005/Atom" href="http://purl.org/net/sword/error/ErrorBadRequest" xmlns:sword="http://purl.org/net/sword/terms/">
  <title>ERROR</title>
  <updated>2022-04-05T18:58:03Z</updated>
  <generator uri="http://www.swordapp.org/" version="2.0"/>
  <summary>problem calling importXML: javax.ejb.EJBException</summary>
  <sword:treatment>Processing failed</sword:treatment>
</sword:error>

[2022-04-05T18:58:03.895+0000] [Payara 5.2021.5] [WARNING] [AS-EJB-00056] [javax.enterprise.ejb.container] [tid: _ThreadID=84 _ThreadName=http-thread-pool::http-listener-1(2)] [timeMillis: 1649185083895] [levelValue: 900] [[
  A system exception occurred during an invocation on EJB ImportGenericServiceBean, method: public void edu.harvard.iq.dataverse.api.imports.ImportGenericServiceBean.importXML(java.lang.String,java.lang.String,edu.harvard.iq.dataverse.DatasetVersion)]]

[2022-04-05T18:58:03.895+0000] [Payara 5.2021.5] [WARNING] [] [javax.enterprise.ejb.container] [tid: _ThreadID=84 _ThreadName=http-thread-pool::http-listener-1(2)] [timeMillis: 1649185083895] [levelValue: 900] [[
  
javax.ejb.EJBException
	at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:723)
	at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:652)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:482)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4592)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2125)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2095)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
	at com.sun.proxy.$Proxy340.importXML(Unknown Source)
	at edu.harvard.iq.dataverse.api.imports.__EJB31_Generated__ImportGenericServiceBean__Intf____Bean__.importXML(Unknown Source)
	at edu.harvard.iq.dataverse.api.datadeposit.CollectionDepositManagerImpl.createNew(CollectionDepositManagerImpl.java:118)
	at org.swordapp.server.CollectionAPI.post(CollectionAPI.java:142)
	at edu.harvard.iq.dataverse.api.datadeposit.SWORDv2CollectionServlet.doPost(SWORDv2CollectionServlet.java:45)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:665)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:750)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:331)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.glassfish.tyrus.servlet.TyrusServletFilter.doFilter(TyrusServletFilter.java:282)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:226)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:257)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:161)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:757)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:577)
	at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:99)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:158)
	at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:371)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:238)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:520)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:217)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:182)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:156)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:218)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:95)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:260)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:177)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:109)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:88)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:53)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:524)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:89)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:94)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:33)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:114)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:569)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:549)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseLicense(JsonParser.java:396)
	at edu.harvard.iq.dataverse.util.json.JsonParser.parseDatasetVersion(JsonParser.java:348)
	at edu.harvard.iq.dataverse.api.imports.ImportGenericServiceBean.importXML(ImportGenericServiceBean.java:106)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:588)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:408)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4826)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:665)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:834)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at jdk.internal.reflect.GeneratedMethodAccessor278.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:615)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:72)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at jdk.internal.reflect.GeneratedMethodAccessor276.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:888)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:833)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:375)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4798)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4786)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	... 44 more
]]

### PR DESCRIPTION
**Pull request by pdurbin**
**What this PR does / why we need it**:

Right now you get a 500 error if you try to create a dataset with SWORD when you have custom terms disabled (:AllowCustomTermsOfUse).

In this pull request, if custom terms are disabled we now report an error if you try to use dcterms:rights.

**Which issue(s) this PR closes**:

- Closes #8580

**Special notes for your reviewer**:

The null pointer was fixed by passing licenseService to JsonParser.

**Suggestions on how to test this**:

Make sure this is true: if custom terms are disabled we now report an error if you try to use dcterms:rights

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No.

**Is there a release notes update needed for this change?**:

No, just a minor bug fix.

**Additional documentation**:

Included (the new error is mentioned).

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/api/datadeposit/SwordServiceBean.java [hunk 1]
```diff
@@ -150,6 +153,13 @@ public void addDatasetSubjectIfMissing(DatasetVersion datasetVersion) {
     public void setDatasetLicenseAndTermsOfUse(DatasetVersion datasetVersionToMutate, SwordEntry swordEntry) throws SwordError {
         Map<String, List<String>> dcterms = swordEntry.getDublinCore();
         List<String> listOfLicensesProvided = dcterms.get("license");
+        List<String> rights = dcterms.get("rights");
+        if (rights != null) {
```
Comment (scolapasta) [inline src/main/java/edu/harvard/iq/dataverse/api/datadeposit/SwordServiceBean.java:157]: very small request, but I think it'd be more "readable" if this was one if statement with an &&.

**Review by scolapasta — CHANGES_REQUESTED** [review]
A very small and specifc change request. Additionally, though, you mention in the issue that we should consider removing the constructor without license server. I agree and since this would be good tech debt clean up (even if it increases the scope). Alternatively, we could have that work be in the fix for the issue I'm about to create from an RT ticket we got. I'm in favor of taking care of that now (in case there are even other circumstances).

File: src/main/java/edu/harvard/iq/dataverse/api/datadeposit/SwordServiceBean.java (hunk 1 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/api/datadeposit/SwordServiceBean.java]: Sure, fixed in 3d4f968

**Comment by pdurbin** [comment]
> Additionally, though, you mention in the issue that we should consider removing the constructor without license server. I agree and since this would be good tech debt clean up (even if it increases the scope). Alternatively, we could have that work be in the fix for the issue I'm about to create from an RT ticket we got.

Yesterday I did a little investigation into removing the old constructor. This pull request removes one instance and there are five more:

Screen Shot 2022-05-10 at 11 55 53 AM

For the bottom two in ImportGenericServiceBean I'd recommend simply deleting both methods where the constructor used. I'm happy to do this in this pull request. It's dead code added in 087b017 before we shipped 4.0 and the app compiles fine without it. One method is an unused variation on importXML and the other is importDCTerms, also unused.

I haven't tried deleting JsonParser with the null, null, null. I guess we can try.

The two in ImportServiceBean give me pause because they're off topic for SWORD. I'm sure doImportHarvestedDataset has to do with harvesting. doImport I played around with a bit but I don't think the import API (/api/batch/import) works at all anymore and it's undocumented. So it would be some effort to get the API working at all before we remove the constructor. And I assume we don't want to document that API. I think it used for DVN 3 migration.

In summary, for this pull request I'm happy to remove the dead code in ImportGenericServiceBean and take a swing at removing the null,null,null JsonParser. Harvesting should probably be handled in a separate effort. /api/batch/import I'm not sure we even want to prioritize fixing.

**Comment by scolapasta** [comment]
@pdurbin ok. let's also mark the method as ```@depecated.```

**Comment by pdurbin** [comment]
@scolapasta sure, deprecated in de9ff67832
